// editorconfig-tools/lib/editorconfigerror.js
// Generated by CoffeeScript 1.8.0

/**
 * The default SyntaxError class doesn't accept fileName or lineNumber args.
 * Also, being able to pass in the context (text that threw the error) is nice.
 */

(function() {
  var EditorConfigError,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  EditorConfigError = (function(_super) {
    __extends(EditorConfigError, _super);

    function EditorConfigError(message, fileName, lineNumber, context) {
      this.message = message;
      this.fileName = fileName;
      this.lineNumber = lineNumber;
      this.context = context;
      this.toString = __bind(this.toString, this);
    }

    EditorConfigError.prototype.toString = function() {
      var text;
      text = this.fileName;
      if (this.lineNumber != null) {
        text += ":" + this.lineNumber;
      }
      if (this.context != null) {
        text += "\n" + context;
      }
      return text += "\nEditorConfigError: " + this.message;
    };

    return EditorConfigError;

  })(Error);

  module.exports = EditorConfigError;

}).call(this);
